Add core types to be used to replace the use of jcr.Session in the kernel API

acoburn opened the pull request:
I have a complete implementation for this, but so as not to make the PR too massive, these are the new interfaces being added which will be used to replace instances of `jcr.Session`, `jcr.Repository`, and the existing `org.fcrepo.kernel.api.services.TransactionService`.

Begins to address: https://jira.duraspace.org/browse/FCREPO-1868

Supersedes #968

ajs6f commented on fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/FedoraSession.java:
```diff
@@ -0,0 +1,150 @@
+
+    public interface State {}
+
+    public enum RequiredState implements State {
```
Wait, what does it mean for a `FedoraSession` to be in `ACTIVE` as opposed to `BATCH`?

acoburn replied:
`ACTIVE` is something that is used in the current impl to test `jcr.Session::isLive`. `BATCH` is a marker to indicate that a particular session persists across multiple HTTP requests.

ajs6f replied:
Do we need to do that here? Can't we just offer 'begin', 'commit', etc., and leave it at that? HTTP notions don't belong here, to my mind.

acoburn replied:
you mean remove `ACTIVE`? I have that here b/c the http layers test for that.

ajs6f replied:
I'm saying I don't see the need for this state machinery at all.

ajs6f commented on fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/FedoraSession.java:
```diff
@@ -0,0 +1,150 @@
+
+    /**
+     * Determine whether the session has the provided state
+     * @param <T> an enum that extends the default State enum
```
You cannot extend a Java `enum`.

acoburn replied:
the comment is wrong -- `State` is an interface, not an `enum`.

ajs6f replied:
Then what is the point of the generic? Doesn't it come to:
```
boolean hasState(State state);
```
?

acoburn replied:
yes, that's correct, I can simplify that signature.

ajs6f commented on fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/FedoraSession.java:
```diff
@@ -0,0 +1,150 @@
+     * @param prefix the prefix
+     * @param uri the URI
+     */
+    void setNamespacePrefix(String prefix, String uri);
```
We don't have any real data-handling methods, but we're still going to demand that you manage namespaces? I don't think this makes sense here.

acoburn replied:
yes, that part can go.

ajs6f commented on fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/FedoraSession.java:
```diff
@@ -0,0 +1,150 @@
+     * @return the default timeout value
+     */
+    public static Duration operationTimeout() {
+       if (System.getProperty(TIMEOUT_SYSTEM_PROPERTY) != null) {
```
How about just`java.lang.System.getProperty(String, String)`?

acoburn replied:
yes, that would be much better

ajs6f commented on fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/services/BatchService.java:
```diff
@@ -0,0 +1,121 @@
+ * @author acoburn
+ * @since Sept 30, 2016
+ */
+public interface BatchService {
```
Is this a kind of "manage the connections between users and open transactions" thing?

acoburn replied:
There is currently an interface called `TransactionService`. This interface is a slight modification of that, mostly to remove some of the baggage of using the word "transaction"

ajs6f replied:
Yeah. I wonder if we want this stuff in the kernel at all.

Partly, this is about the fact that I don't think we have any clear sense of why there are all these `*Service` guys.

acoburn replied:
The reason they are in the kernel is so that the HTTP layers can access these "services" via spring injection. Moving those out would be a structural change that goes way beyond the scope of this effort.

ajs6f replied:
No, I'm not asking you to do that. I just don't want to make any effort to preserve them if we can avoid it.

acoburn replied:
Yes, but thinking about a possible OSGi-based impl, these "service" interfaces could come in really handy.

ajs6f replied:
Okay, leave it be for now. I'm certainly not trying to block this.

acoburn replied:
@ajs6f: related to the `org.fcrepo.kernel.api.services.*Service` types, I wonder whether they might not be better situated in a different package -- say: `org.fcrepo.spi` or `org.fcrepo.kernel.spi` (which could be independent of the `fcrepo-kernel-api` module). Once we have a full specification and remove all of the JCR concepts from the API, I think there's a certain amount of re-organization that can happen w/r/t the modules and package namespaces, but that's a conversation for later.

ajs6f replied:
Yeah, let's get through this. I have yet to hear anyone give a _purpose_ for those types. (I know you have OSGi in mind, but I don't think could have been our original purpose. If we decide to have these kinds of types for OSGi, they should be impl'd in a common OSGi-related module, not in the kernel-api forcing everyone to impl them directly.)

acoburn replied:
I'd argue that the _purpose_ of these types is to make it possible for the HTTP layer (fcrepo-http-api and fcrepo-http-commons) to access those implementation-specific services that are injected via Spring (but that could be substituted with Blueprint for OSGi). Either way, I agree that they are not a necessary part of the kernel-api.

ajs6f replied:
They do that, but they are only _needed_ to do that on the assumption that model types (like `FedoraResource`) aren't managed by the bean container or other outer system. @awoods and I have talked in the past about whether that's a step we ought to contemplate taking. My point here is that whatever we do in the JCR impl, other people may want to play it out differently elsewhere and we shouldn't make their lives any harder.

awoods commented on fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/services/BatchService.java:
```diff
@@ -0,0 +1,121 @@
+     * @param session The session to use for this batch operation
+     * @param username the name of the {@link java.security.Principal}
+     */
+    void begin(FedoraSession session, String username);
```
From the FedoraSession interface, it would appear that a `userId` is already associated with sessions. How is this arg `username` reconciled with the session userId?

acoburn replied:
You are correct: the jcr.Session interface has a getUserID method, and the new FedoraSession interface has a getUserId method -- which in the impl just proxies to the jcr.Session::getUserID. The issue is that in the case of an anonymous user, the result of the jcr.Session::getUserID is not `null`, but rather "<anonymous>". But we can't actually rely on that string, because that, too, is configurable in the repository.json file. This way, we are relying on the value of the java.security.Principal::getName value for looking up FedoraSession values.

Believe me -- my first stab at this excluded the `username` parameter on all of these methods, but relying on that conflicts with how transaction values are looked up.

awoods commented on fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/services/BatchService.java:
```diff
@@ -0,0 +1,121 @@
+     * @param username the name of the {@link java.security.Principal}
+     * @return the {@link FedoraSession} with this user
+     */
+    FedoraSession getSession(String txId, String username);
```
What is the expected result if the `txId` exists but not with the provided arg `username`?

acoburn replied:
I did not change the behavior of the existing TransactionServiceImpl, namely an Exception is thrown: https://github.com/fcrepo4/fcrepo4/blob/master/fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/services/TransactionServiceImpl.java#L137-L138

Personally, I'd prefer to return an Optional<FedoraSession>, but this PR is about removing the jcr.Session, not about changing everything.

awoods commented on fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/services/BatchService.java:
```diff
@@ -0,0 +1,121 @@
+     * @param username the name of the {@link java.security.Principal}
+     * @return the {@link FedoraSession} object for the defined user
+     */
+    boolean exists(String txid, String username);
```
Minor: change `txid` to `txId` for consistency.
..or the other way around.

acoburn replied:
I'd go further and remove the "tx" part altogether, just using `sessionId`.

acoburn replied:
@awoods I changed this and all subsequent `txid` names to `sessionId`.

awoods commented on fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/services/BatchService.java:
```diff
@@ -0,0 +1,121 @@
+     * @param txid the Id of the {@link FedoraSession}
+     * @return the {@link FedoraSession} object
+     */
+    default boolean exists(String txid) {
```
Minor: change `txid` to `txId` for consistency.
..or the other way around.

awoods commented on fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/services/BatchService.java:
```diff
@@ -0,0 +1,121 @@
+     * @param txid the id of the {@link FedoraSession}
+     * @param username the name of the {@link java.security.Principal}
+     */
+    void commit(String txid, String username);
```
Minor: change `txid` to `txId` for consistency.
..or the other way around.

awoods commented on fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/services/BatchService.java:
```diff
@@ -0,0 +1,121 @@
+     *
+     * @param txid the id of the {@link FedoraSession}
+     */
+    default void commit(String txid) {
```
...

awoods commented on fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/services/BatchService.java:
```diff
@@ -0,0 +1,121 @@
+     * @param txid the id of the {@link FedoraSession}
+     * @param username the name of the {@link java.security.Principal}
+     */
+    void abort(String txid, String username);
```
...

awoods commented on fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/services/BatchService.java:
```diff
@@ -0,0 +1,121 @@
+     *
+     * @param txid the id of the {@link FedoraSession}
+     */
+    default void abort(String txid) {
```
...

awoods commented:
:+1:

acoburn commented:
@awoods, @ajs6f I added the modeshape implementation for these new interfaces. Once I get the green light on this, I'll add the rest of the commits (which actually remove the use of jcr.Session from the kernel-api).

ajs6f commented on fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/FedoraRepositoryImpl.java:
```diff
@@ -0,0 +1,87 @@
+     *
+     * @return the JCR Repository
+     */
+    public Repository getRepository() {
```
This naming is confusing; maybe `getJcrRepository`?

acoburn replied:
That works for me. Also, I should double check the use of this method, because we may be able to do away with this altogether.

acoburn replied:
@ajs6f I renamed the method here and in `FedoraSessionImpl` (`getSession` -> `getJcrSession`). The `FedoraRepositoryImpl::getJcrRepository` method is only used for unit tests -- I don't know if that's a good reason to keep it, but generally, I tried to keep existing unit/integration tests in place. All things being equal, I would probably remove that test and therefore that method, too. @awoods: thoughts?

acoburn commented:
@awoods @ajs6f let me know when you are ready for the full implementation. Barring any changes with the code so far, the next commit will be `76 files changed, 1019 insertions(+), 2175 deletions(-)`

ajs6f commented on fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/FedoraSession.java:
```diff
@@ -0,0 +1,92 @@
+    Instant getCreated();
+
+    /**
+     * Get the date this session expires
```
We've talked about immortal sessions. Either there should be a note explaining what to use for that "value" or this should be `Optional<Instant>` (I like the latter).

acoburn replied:
I updated the FedoraSession interface to return an `Optional<Instant>`

ajs6f commented on fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/FedoraSession.java:
```diff
@@ -0,0 +1,92 @@
+     * @param key the key
+     * @param value the value
+     */
+    void setSessionData(String key, String value);
```
What about multi-valued keys?

ajs6f commented on fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/FedoraRepositoryImpl.java:
```diff
@@ -0,0 +1,87 @@
+                throw new RepositoryRuntimeException(ex);
+            }
+        }
+        throw new IllegalArgumentException("Credentials are of the wrong type");
```
Can this be a more specific subtype?

ajs6f commented on fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/FedoraRepositoryImpl.java:
```diff
@@ -0,0 +1,87 @@
+        if (repository instanceof FedoraRepositoryImpl) {
+            return ((FedoraRepositoryImpl)repository).getJcrRepository();
+        }
+        throw new IllegalArgumentException("FedoraRepository is of the wrong type");
```
Ditto-- maybe a more specific exception?

ajs6f commented on fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/FedoraSessionImpl.java:
```diff
@@ -0,0 +1,182 @@
+
+    public static final String TIMEOUT_SYSTEM_PROPERTY = "fcrepo.session.timeout";
+
+    private final Session session;
```
This name is a little confusing: maybe call it `jcrSession` or something?

ajs6f commented on fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/FedoraSessionImpl.java:
```diff
@@ -0,0 +1,182 @@
+ * An implementation of the FedoraSession abstraction
+ * @author acoburn
+ */
+public class FedoraSessionImpl implements FedoraSession {
```
If this class is not thread-safe (and it does not appear to be) we should note that in the doc.

ajs6f commented on fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/FedoraSessionImpl.java:
```diff
@@ -0,0 +1,182 @@
+        if (session instanceof FedoraSessionImpl) {
+            return ((FedoraSessionImpl)session).getJcrSession();
+        }
+        throw new IllegalArgumentException("FedoraSession is of the wrong type");
```
Maybe a more specific exception type?

ajs6f commented on fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/services/BatchServiceImpl.java:
```diff
@@ -0,0 +1,135 @@
+
+    @Override
+    public FedoraSession getSession(final String sessionId, final String username) {
+        final FedoraSession tx = transactions.computeIfAbsent(getTxKey(sessionId, username), s -> {
```
I like Java 8 but this seems a bit fancy-shmancy. How about
```
String txKey = getTxKey(sessionId, username);
if (!transactions.containsKey(txKey)) {
    throw new SessionMissingException("Batch session with id: " + txKey + " is not available!");
}
return transactions.get(txKey);
```

ajs6f replied:
The Javadocs for `Map::computeIfAbsent` say `The default implementation makes no guarantees about synchronization or atomicity properties of this method.` so you are relying on `ConcurrentHashMap`s impl-specific behavior here.... but do we have any understanding anyway of the atomicity of `*Service`? I don't think this is coherent.

ajs6f commented on fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/services/BatchServiceImpl.java:
```diff
@@ -0,0 +1,135 @@
+
+    @Override
+    public void commit(final String sessionId, final String username) {
+        final FedoraSession session = transactions.remove(getTxKey(sessionId, username));
```
If the `session.commit()` fails, don't you not want to have removed it from the transactions? The client might want to retry.

ajs6f commented on fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/services/BatchServiceImpl.java:
```diff
@@ -0,0 +1,135 @@
+
+    @Override
+    public void abort(final String sessionId, final String username) {
+        final FedoraSession tx = transactions.remove(getTxKey(sessionId, username));
```
Same as above-- if the expire fails, shouldn't the session still be available for further detective work?

ajs6f commented on fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/services/BatchServiceImpl.java:
```diff
@@ -0,0 +1,135 @@
+    }
+
+    private static String getTxKey(final String sessionId, final String username) {
+        return username == null ? sessionId : username + ":" + sessionId;
```
Maybe it's just me, but I would actually prefer anonymous sessions to have the colon.

acoburn commented:
@ajs6f I addressed the latest round of comments in the latest commit.

ajs6f commented on fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/FedoraSession.java:
```diff
@@ -77,16 +80,31 @@
 
     /**
-     * Set session-specific data
+     * Add session-specific data
```
Can we have a note here that it is up to the impl whether multiple values per key are afforded, and a note on the immediate impl that they aren't?

ajs6f commented on fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/FedoraSessionImpl.java:
```diff
@@ -53,10 +55,10 @@
     private final String id;
     private final Instant created;
-    private final Map<String, String> sessionData;
+    private final ConcurrentHashMap<String, String> sessionData;
```
If we aren't claiming that this is thread-safe, what's the point of `ConcurrentHashMap`?

ajs6f replied:
I don't get that. If multiple threads are going to hit a single instance of this class, how is it okay for this class to not be thread-safe?

ajs6f replied:
https://docs.jboss.org/author/display/MODE/Repository+and+Session?_sscc=t#RepositoryandSession-Threadsafety

Grumble...

ajs6f replied:
I don't think we should make guarantees, and if we don't, we should use the cheapest `Map` that does the job.

ajs6f commented on fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/services/BatchServiceImpl.java:
```diff
@@ -92,10 +93,13 @@ public void begin(final FedoraSession session, final String username) {
-            throw new SessionMissingException("Batch session with id: " + s + " is not available");
-        });
-        return tx;
+        System.out.println(transactions.keySet().toString());
```
Looks like some debugging juice?

acoburn replied:
Ooops!

acoburn commented:
@ajs6f more comments addressed in the latest commit.

awoods commented on fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/services/BatchService.java:
```diff
@@ -0,0 +1,138 @@
+    void begin(FedoraSession session, String username);
+
+    /**
+     * Create a new FedoraSession for the anonymous user and add it to the currently open ones
```
I am confused by: "Create a new FedoraSession..." since a `FedoraSession` is being passed into this method.

awoods commented on fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/services/BatchService.java:
```diff
@@ -0,0 +1,138 @@
+     * @param username the name of the {@link java.security.Principal}
+     * @return the {@link FedoraSession} with this user
+     */
+    FedoraSession getSession(String sessionId, String username);
```
What happens if no such session for the provided username exists?

acoburn replied:
A `SessionMissingException` is thrown.

awoods commented on fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/services/BatchService.java:
```diff
@@ -0,0 +1,138 @@
+    }
+
+    /**
+     * Refresh an existing session using an implementation-defined default
```
What is the difference between "Refresh an existing session" here and `FedoraSession.updateExpiry`? https://github.com/fcrepo4/fcrepo4/pull/1107/files#diff-2a974e96c1ebe0bcd554be7d704b6955R48

acoburn replied:
The BatchService implementation will typically involve an in-memory (possibly distributed) map of FedoraSessions. `BatchService::refresh` will update the expiry of one of those batch sessions. I.e., BatchService::refresh will probably call FedoraSession::updateExpiry, but as for how that is done is up to the implementation.

awoods commented on fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/services/BatchService.java:
```diff
@@ -0,0 +1,138 @@
+     * @param sessionId the id of the {@link FedoraSession}
+     * @param username the name of the {@link java.security.Principal}
+     */
+    void commit(String sessionId, String username);
```
There seems to be some overlap between this interface and `FedoraSession`. How are we reconciling this commit with `FedoraSession.commit`?

awoods commented on fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/services/BatchServiceImpl.java:
```diff
@@ -0,0 +1,133 @@
+     */
+    private static Map<String, FedoraSession> transactions = new ConcurrentHashMap<>();
+
+    public static final long REAP_INTERVAL = 1000;
```
This can be `private`.

acoburn replied:
this is public (I believe) to facilitate testing. I can add the appropriate guava annotation.

awoods commented on fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/services/BatchServiceImpl.java:
```diff
@@ -0,0 +1,133 @@
+     * be either persisted or written to a distributed map or sth, not just this
+     * plain hashmap that follows
+     */
+    private static Map<String, FedoraSession> transactions = new ConcurrentHashMap<>();
```
Would `sessions` be a clearer name for this map? It seems like we are otherwise trying to get away from the term  "transaction".

acoburn replied:
sure

awoods commented on fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/FedoraSessionImpl.java:
```diff
@@ -0,0 +1,205 @@
+public class FedoraSessionImpl implements FedoraSession {
+
+    // The default timeout is 3 minutes
+    public static final String DEFAULT_TIMEOUT = Long.toString(ofMinutes(3).toMillis());
```
Can be `private`.

acoburn replied:
same as above...

awoods commented on fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/FedoraSessionImpl.java:
```diff
@@ -0,0 +1,205 @@
+    // The default timeout is 3 minutes
+    public static final String DEFAULT_TIMEOUT = Long.toString(ofMinutes(3).toMillis());
+
+    public static final String TIMEOUT_SYSTEM_PROPERTY = "fcrepo.session.timeout";
```
Can be `private`.

acoburn replied:
same as above...

awoods commented on fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/FedoraSessionImpl.java:
```diff
@@ -0,0 +1,205 @@
+    private static final ObjectMapper mapper = new ObjectMapper();
+
+    /**
+     * Create a Fedora session with a JCR session and a username
```
This javadoc should probably be updated to remove reference to `username`.

awoods commented on fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/FedoraSessionImpl.java:
```diff
@@ -0,0 +1,205 @@
+     *  @param key the data key
+     *  @param value the data value
+     *
+     *  Note: while the FedoraSession inteface permits multi-valued
```
Typos: `inteface` and `assciated`

awoods replied:
Have you verified that the current codebase does not require appending values to an existing UserData key?

acoburn replied:
yes

awoods commented on fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/FedoraSessionImpl.java:
```diff
@@ -51,8 +52,10 @@
 
     // The default timeout is 3 minutes
+    @VisibleForTesting
     public static final String DEFAULT_TIMEOUT = Long.toString(ofMinutes(3).toMillis());
```
This is not actually used in any testing code... and can presently be safely `private`.

awoods commented on fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/FedoraSessionImpl.java:
```diff
@@ -51,8 +52,10 @@
+    @VisibleForTesting
     public static final String DEFAULT_TIMEOUT = Long.toString(ofMinutes(3).toMillis());
 
+    @VisibleForTesting
```
This is not actually used in any testing code... and can presently be safely `private`.

awoods commented on fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/services/BatchServiceImpl.java:
```diff
@@ -60,8 +61,9 @@
-    private static Map<String, FedoraSession> transactions = new ConcurrentHashMap<>();
+    private static Map<String, FedoraSession> sessions = new ConcurrentHashMap<>();
 
+    @VisibleForTesting
```
This is not actually used in any testing code... and can presently be safely `private`.

awoods commented:
This is ready to go from my perspective. @ajs6f, have all of your comments been addressed?

ajs6f commented:
@awoods I have no idea. That's not @acoburn's fault at all-- I find this new Github "review" mode terribly confusing and unhelpful and I can no longer keep track of this conversation. Frankly, if @acoburn gives us his word that he addressed everything, that's fine with me.

acoburn commented:
And I find that this "review mode" makes it difficult to keep track of issues that I've resolved. As far as I know, I have addressed all of the issues brought up in this discussion, but it has also involved a lot of back and forth, and it is entirely possible that something has fallen between the cracks.

awoods commented:
It sounds like, to the best of our collective knowledge, this PR is ready to go. Feel free to merge it @ajs6f .

ajs6f commented:
Am I to close and reopen the ticket?

acoburn commented:
@ajs6f just reopen the Jira ticket.
